docs/DafnyRef/UserGuide.md

```diff
@@ -156,14 +156,227 @@ exit code.
+* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
+* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
+* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.
+* Every [`decreases` clause](#sec-decreases-clause) yields an _assertion_ at either a call site or at the end of a while loop.
+* Every non-empty [subset type](#sec-subset-types) yields an _assertion_ that its witness satisfies the constraint.
+
+[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.
+
+It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],
+except for `ensures` assertions that appear earlier in the code but, for verification purposes, would appear at the end.
+Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,
+and 2) every preceding assumption (and previous assertions transformed in assumptions).
+
+[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.
+
+To reduce overhead, Dafny collects all the assertions of one method into one single conjecture that it sends to the verifier, which tries to prove it correct:
+
+* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
+* If the verifier returns a counterexample, this counter-example is used to determine both the failing assertion and the failing path.
+  In order to retrieve additional failing assertions, Dafny will again query the verifier after turning previously failed assertions into assumptions, [like this](https://github.com/dafny-lang/dafny/discussions/1898#discussioncomment-2344533).[^caveat-about-assertion-and-assumption]
+* If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny partitions the assertions up into smaller batches[^smaller-batches]. An extreme case is the use of the `/vcsSplitOnEveryAssert` command-line option or the [`{:vcs_split_on_every_assert}` attribute](#sec-vcs_split_on_every_assert), which causes Dafny to make one batch for each assertion.
```

Comment:
Just for my own sanity: I know Boogie has options such as `/vcsMaxKeepGoingSplits` that cause this adaptive behavior, but my understanding is that they are not on by default, and that is critical to make sure verification is deterministic.

Reply:
Correct.

Reply:
Okay, then we need to clarify that this bullet isn't true by default.